TOF channel calib implemented as workflow

chiarazampolli opened the pull request.

chiarazampolli committed:
First version

Not tested, it is Friday evening...

Adding to compilation

Not working yet, still using a struct for TOFChannelData, but with the boost
histogram it seems too complicated. Now I will try using a class. I make
the commit only to save the current version.

Compiling TOFChannelData for now

Adding TOFChannelCalibrator to compilation - but it fails

adding destructors

Fixes + updates

Using 18 histograms

Trying to make integral work - not successful for now

misc fixes (#18)

Co-authored-by: shahoian <ruben.shahoyan@cern.ch>

Fixing integral function

Leaving for now alternative methods, just for reference (they
do not work as wanted anyway)

Better integral functions.

In the functions, there are still the three possibilities:
1) integral over reduced indexed loop (fastest);
2) integral over the complete indexed loop;
3) projection + accumulate (slowest).

They are kept for reference for now, until the boost people
find maybe a better way (see https://github.com/boostorg/histogram/issues/277)

Improve finalizeSlot

Adding faster way to get the integral

shahor02 committed:
Fix in calibration slots queue creation

+ define TFType only once in the TimeSlot.
+ rename TimeSlotCalibration::mTFStart to mFirstTF, to avoid confusion with
  TimeSlot::mTFStart

chiarazampolli committed:
Workflow for TOFChannelCalibration

But it does not compile...

chiarazampolli committed:
Version that does the calibration

Fix after rebase

not sending output at every TF (and typo fixed)

chiarazampolli committed:
Reading existing CCDB implemented.

This is not tested yet.
What is not clear, also, is how to retrieve the timestamp of the current TF,
which is needed to verify if the latest calib entry available is too old.
For now I just used a "placeholder".
Also the startTimeValidity of the CCDB object should be passed by the
framework. Now I am still using the dummy workflow (tof-dummy-ccdb).

fixes, but still I get a crash - this commit is to share with others

Next try... Still not working

Now working

chiarazampolli committed:
Changing the type of mTimeSlewing

chiarazampolli committed:
Changing type of the elements in the pairs of the TimeSlewing vector

From <float, float> to <unsigned short, short>

chiarazampolli committed:
One calib workflow for both LHCphase and TOF channel offsets

and possibility to use the previously calculated LHCphase
during the TOFchannel offset calibration. Most of this is
just a comment for now, to set the idea on how to do it.

Fix and protection.

- fix for retriving offset from TimeSlewing object;
- protection for possible division by 0

Adapting to generate LHCphase consisent with few ns and other stuff.

Some are debug printouts that will be removed.

Some fixes

Removed useless lines, comments, change LOG(INFO) --> LOG(DEBUG)...

clang-format

chiarazampolli committed:
Fixing check for TS object + new method for TS object

chiarazampolli committed:
Creating a tof-dummy-ccdb-for-calib

This is a workflow for calibration, where I don't have the lines:

control.endOfStream();
control.readyToQuit(QuitRequest::Me);

which prevents it from running at every time frame. In addition,
the tof-dummy-ccdb-for-calib is added as executable, not as workflow,
which avoids the message during compilation of:

Invalid workflow: No matching output found for DUM/CALIBDATA/0. Candidates:
-TOF/ChannelCalib/0
-TOF/LHCphase/0
-TOF/StartChCalib/0
-TOF/StartLHCphase/0

clang-format

chiarazampolli committed:
forgotten return value

clang-format...

Following Ruben's review

clang-format

chiarazampolli commented:
Hello,
This PR is failing in a test:

286/318 Test #276: o2sim_G3 .......................................................................***Failed  Required regular expression not found. Regex=[SIMULATION RETURNED SUCCESFULLY
]301.64 sec

I also see:

Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libRIO, output: $ROOT_DYN_PATH/libRIO
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libCling, output: $ROOT_DYN_PATH/libCling
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Test attempt 1/3 failed with exit code 124 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Running /mnt/mesos/sandbox/sandbox/sw/BUILD/5c131a43d8c8b6caf4557baee4af770057f8b88f/O2/stage/tests/o2-test-tpc-FastTransform with args  (attempt 2/3) ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Process running as 50250 for attempt 2 ===
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libRIO, output: $ROOT_DYN_PATH/libRIO
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libCling, output: $ROOT_DYN_PATH/libCling
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Test attempt 2/3 failed with exit code 124 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Stack trace follows for attempt 2 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - End of stack trace for attempt 2 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Running /mnt/mesos/sandbox/sandbox/sw/BUILD/5c131a43d8c8b6caf4557baee4af770057f8b88f/O2/stage/tests/o2-test-tpc-FastTransform with args  (attempt 3/3) ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Process running as 52286 for attempt 3 ===
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libRIO, output: $ROOT_DYN_PATH/libRIO
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libCling, output: $ROOT_DYN_PATH/libCling
Running 2 test cases...
average difference in correction 1.06558e-05 cm 
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Test attempt 3/3 failed with exit code 124 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Stack trace follows for attempt 3 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - End of stack trace for attempt 3 ===
=== Detectors/TPC/reconstruction/test/testTPCFastTransform.cxx - Test failed after 3 attempts with 124 ===

Any idea? @sawenzel ? It was not failing when @noferini approved it (which was not enough to merge it, apparently).

shahor02 commented:
The failure is indeed unrelated. I'll check the PR and merge if it is ok later today.

chiarazampolli commented:
Thanks!

shahor02 commented on DataFormats/Detectors/TOF/include/DataFormatsTOF/CalibTimeSlewingParamTOF.h:
```diff
@@ -73,8 +80,8 @@ class CalibTimeSlewingParamTOF
-  std::vector<std::pair<float, float>>* mTimeSlewing[18]; //! pointers to the sector vectors
-
+  std::vector<std::pair<unsigned short, short>> mTimeSlewing[18]; ///< array of sector vectors; first element of the pair is TOT (in ps), second is t-texp_pi (in ps)
+  /*
```
Since these commented line will never be used, could you suppress them?

chiarazampolli replied:
Done

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -40,7 +40,7 @@ CalibTimeSlewingParamTOF::CalibTimeSlewingParamTOF()
   mTimeSlewing[16] = &mTimeSlewingSec16;
   mTimeSlewing[17] = &mTimeSlewingSec17;
-
+  */
```
also here

chiarazampolli replied:
Done

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -50,11 +50,66 @@ CalibTimeSlewingParamTOF::CalibTimeSlewingParamTOF()
   }
 }
 //______________________________________________
```
with the huge number of channels you have and a special value -1 it is much faster to do
````
for (int i = 0; i < NSECTORS; i++) {
  memset(mChannelStart[i],-1,sizeof(mChannelStart[i][0]));
}
````

chiarazampolli replied:
I guess you mean:

`memset(mChannelStart[i],-1,sizeof(mChannelStart[i]));`

chiarazampolli replied:
Done

shahor02 replied:
actually, I meant ``memset(mChannelStart[i],-1,NCHANNELXSECTOR*sizeof(mChannelStart[i][0]));`` but for the fixed size array also your line will work.

chiarazampolli replied:
Yes, they are equivalent.

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -50,11 +50,66 @@ CalibTimeSlewingParamTOF::CalibTimeSlewingParamTOF()
+  for (int iSec = 0; iSec < NSECTORS; iSec++) {
+    mTimeSlewing[iSec] = source.mTimeSlewing[iSec];
+  }
+  /*
```
this commented code also is not needed.

chiarazampolli replied:
Removed

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -50,11 +50,66 @@ CalibTimeSlewingParamTOF::CalibTimeSlewingParamTOF()
+  */
+  for (int i = 0; i < NSECTORS; i++) {
+    for (int j = 0; j < NCHANNELXSECTOR; j++) {
+      mChannelStart[i][j] = source.mChannelStart[i][j];
```
here too, these method will be used at every relocation and instead of huge loop you can do just 1 copy
````
memcpy(mChannelStart[i], source.mChannelStart[i], NCHANNELXSECTOR*sizeof(mChannelStart[i][0]));
memcpy(mFractionUnderPeak[i], source.mFractionUnderPeak[i], NCHANNELXSECTOR*sizeof(mFractionUnderPeak[i][0]));
memcpy(mSigmaPeak[i], source.mSigmaPeak[i], NCHANNELXSECTOR*sizeof(mSigmaPeak));
````

chiarazampolli replied:
I guess you mean:
```
memcpy(mChannelStart[i], source.mChannelStart[i], sizeof(mChannelStart[i]));
    memcpy(mFractionUnderPeak[i], source.mFractionUnderPeak[i], sizeof(mFractionUnderPeak[i]));
    memcpy(mSigmaPeak[i], source.mSigmaPeak[i], sizeof(mSigmaPeak[i]));
```

right?

shahor02 replied:
no, but your lines also will work (and yes, I had a type in the 3d line)

chiarazampolli replied:
Yes, also here they are equivalent.

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -50,11 +50,66 @@ CalibTimeSlewingParamTOF::CalibTimeSlewingParamTOF()
+    mTimeSlewing[i] = source.mTimeSlewing[i];
 
     for (int j = 0; j < NCHANNELXSECTOR; j++) {
       mChannelStart[i][j] = source.mChannelStart[i][j];
```
idem, prefer copy

chiarazampolli replied:
Done, but please check the comment above.

shahor02 commented on Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:
```diff
@@ -0,0 +1,521 @@
+
+  return res2;
+
+  /* // what is below is only for alternative methods which all proved to be slower
```
if you are sure this will be never used, better to suppress it.

chiarazampolli replied:
I would prefer to keep it if it is not against the rules.

shahor02 commented on Detectors/TOF/prototyping/ConvertRun2CalibrationToO2.C:
```diff
@@ -76,12 +77,17 @@ void ConvertRun2CalibrationToO2()
+  Int_t nChannelToCheck = 0;
+  int forced = 0;
+  int sizeRecalc = 0;
   for (Int_t i = 0; i < 157248; i++) {
```
use defined constexpr?

chiarazampolli replied:
Replaced with AliTOFGeometry functions.

shahor02 commented on Detectors/TOF/prototyping/ConvertRun2CalibrationToO2.C:
```diff
@@ -76,12 +77,17 @@ void ConvertRun2CalibrationToO2()
 
   Float_t x[10000];
```
what defines these 10000? Here you allocate 160KB on the stack. Normally should not be critical, but is it needed?

chiarazampolli replied:
This is a macro used to create O2 CCDB-like objects from Run2-OCDB objects. It does not need to be optimized. 10000 is simply to be sure to have enough. I would leave it as it is.

shahor02 replied:
ok, sure

shahor02 reviewed (commented):
Hi @chiarazampolli please see a few comments below.

chiarazampolli commented:
Hello @shahor02 ,
I addressed all your comments. Only the one on the Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx file I did not take: i prefer to leave the commented out lines, for reference, if possible.
Thanks for reviewing!
Chiara

shahor02 commented on DataFormats/Detectors/TOF/src/CalibTimeSlewingParamTOF.cxx:
```diff
@@ -20,53 +20,52 @@ using namespace o2::dataformats;
+{
+  for (int i = 0; i < NSECTORS; i++) {
+    memset(mChannelStart[i], -1, sizeof(mChannelStart[i]));
+    memset(mFractionUnderPeak[i], -100, sizeof(mFractionUnderPeak[i]));
```
Hi @chiarazampolli, sorry, I've just noticed that the ``mSigmaPeak`` and ``mFractionUnderPeak`` are floats, for them the memset -1 will not work (and -100 will not work even for int, I have to change my glasses). 
I can fix it myself but to be sure, why do you need for them -1 and -100 to signal anomaly? Would not 0 work? 
Per-element initialization of such a large arrays is really quite time consuming.

chiarazampolli replied:
See https://github.com/chiarazampolli/AliceO2/pull/19

shahor02 committed:
small fixes (#19)

* small fixes

* Removing commented lines, as discussed in PR

Co-authored-by: shahoian <ruben.shahoyan@cern.ch>
Co-authored-by: chiarazampolli <chiara.zampolli@cern.ch>

clang-format again...

shahor02 commented:
could you pass it by clang-format?

chiarazampolli commented:
Sure, i usually do, but I forgot yesterday. Will do it later today.

shahor02 commented:
@chiarazampolli shall I simply squash it to 1 commit with log = title?

chiarazampolli commented:
Hello @shahor02 , 

Could we leave the commits as separate? Especially some, I prefer that they are clearly divided by others. I already squashed what I thought was ok to squash. If it is a problem, I can squash more, but I would prefer not all.

Thanks!

Chiara

shahor02 commented:
Ok, will merge once the CI has passed
